loki/transform/transform_remove_code.py

```diff
@@ -0,0 +1,333 @@
+
+    def __init__(
+            self, remove_marked_regions=True, mark_with_comment=True,
+            remove_dead_code=False, use_simplify=True,
+            call_names=None, intrinsic_names=None,
+            remove_imports=True, kernel_only=False
+    ):
+        self.remove_marked_regions = remove_marked_regions
+        self.mark_with_comment = mark_with_comment
+
+        self.remove_dead_code = remove_dead_code
+        self.use_simplify = use_simplify
+
+        self.call_names = as_tuple(call_names)
+        self.intrinsic_names = as_tuple(intrinsic_names)
+        self.remove_imports = remove_imports
+
+        self.kernel_only = kernel_only
+
+    def transform_subroutine(self, routine, **kwargs):
+
+        if self.kernel_only and not kwargs.get('role') == 'kernel':
+            return
+
+        # Apply named node removal to strip specific calls
+        if self.call_names or self.intrinsic_names:
+            do_remove_calls(
+                routine, call_names=self.call_names,
+                intrinsic_names=self.intrinsic_names,
+                remove_imports=self.remove_imports
+            )
+
+        # Apply marked region removal
+        if self.remove_marked_regions:
+            do_remove_marked_regions(
+                routine, mark_with_comment=self.mark_with_comment
+            )
+
+        # Apply Dead Code Elimination
+        if self.remove_dead_code:
+            do_remove_dead_code(routine, use_simplify=self.use_simplify)
+
+
+def do_remove_dead_code(routine, use_simplify=True):
+    """
+    Perform Dead Code Elimination on the given :any:`Subroutine` object.
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine to which to apply dead code elimination.
+    simplify : boolean
+        Use :any:`simplify` when evaluating expressions for branch pruning.
+    """
+
+    transformer = RemoveDeadCodeTransformer(use_simplify=use_simplify)
+    routine.body = transformer.visit(routine.body)
+
+
+class RemoveDeadCodeTransformer(Transformer):
+    """
+    :any:`Transformer` class that removes provably unreachable code paths.
+
+    The pirmary modification performed is to prune individual code branches
+    under :any:`Conditional` nodes.
+
+    Parameters
+    ----------
+    use_simplify : boolean
+        Use :any:`simplify` when evaluating expressions for branch pruning.
+    """
+
+    def __init__(self, use_simplify=True, **kwargs):
+        super().__init__(**kwargs)
+        self.use_simplify = use_simplify
+
+    def visit_Conditional(self, o, **kwargs):
+        condition = self.visit(o.condition, **kwargs)
+        body = as_tuple(flatten(as_tuple(self.visit(o.body, **kwargs))))
+        else_body = as_tuple(flatten(as_tuple(self.visit(o.else_body, **kwargs))))
+
+        if self.use_simplify:
+            condition = simplify(condition)
+
+        if condition == 'True':
+            return body
+
+        if condition == 'False':
+            return else_body
```

Comment:
I realise that this has just been moved from the old module, but this comparison against a string put me off for a second... This is using the implicit String comparison to evaluate the condition? Would we want to account for something like `IF(1)`?

Reply:
Yes, this is actually intentional, as `IF(1)` is basically illegal in a Fortran-context (strong typing; `1` is not a `logical`) The visual oddness comes from `"True"` actually being our internal representation of `.true.` (the fortran intrinsic logical value), and that is the only thing we should match in a condition, since everything else should evaluate to that via `simplify` (eg. result of a logical operator will be a logical, never a numerical value).